contracts/governance/FeeSharingProxy.sol

```diff
@@ -69,11 +81,53 @@ contract FeeSharingProxy is SafeMath96, IFeeSharingProxy {
         uint poolTokenAmount = ILoanToken(loanPoolToken).mint(address(this), amount);
         _writeTokenCheckpoint(loanPoolToken, safe96(poolTokenAmount, "FeeSharingProxy::withdrawFees: pool token amount exceeds 96 bits"));
 
-        lastFeeWithdrawalTime = block.timestamp;
+        lastFeeWithdrawalTime[_token] = block.timestamp;
         
         emit FeeWithdrawn(msg.sender, loanPoolToken, poolTokenAmount);
     }
 
+    /**
+     * @notice transfer tokens to this contract
+     * @dev we just update amount of tokens here and write checkpoint in a separate methods
+     * in order to prevent adding checkpoints too often
+     * @param _token address of the token
+     * @param _amount amount to be transferred
+     * */
+    function transferTokens(address _token, uint96 _amount) public {
+        require(_token != address(0), "FeeSharingProxy::transferTokens: invalid address");
+        require(_amount > 0, "FeeSharingProxy::transferTokens: invalid amount");
+
+        //transfer tokens from msg.sender
+        bool success = IERC20(_token).transferFrom(address(msg.sender), address(this), _amount);
+        require(success, "Staking::transferTokens: token transfer failed");
+
+        //update unprocessed amount of tokens
+        unprocessedAmount[_token] = add96(unprocessedAmount[_token], _amount, "FeeSharingProxy::transferTokens: amount exceeds 96 bits");
+
+        emit TokensTransferred(msg.sender, _token, _amount);
+    }
+
+    /**
+     * @notice adds checkpoint with accumulated amount by function invocation
+     * @param _token address of the token
+     * */
+    function addCheckpoint(address _token) public {
```

Comment:
why not internal?

(no full review at this point, just quick scan)

Reply:
- `transferTokens` will be invoked by Staking contract for all early unstaking
- `addCheckpoint` should be invoked externally in order to create checkpoint (currently, we allow to create new checkpoint after 24h after previous one), we need to think about scheduling it
- `withdraw` will be invoked by user, all newly created and unprocessed checkpoint should be processed here

Reply:
why do we need to add a checkpoint externally?
don't we only need it after withdrawing fees and transferring tokens?

Reply:
We don't want to create a checkpoint for every slashing of tokens.
I think it's better to invoke `addCheckpoint` from `transferTokens` if 24h passed since last checkpoint.
Can change it tomorrow.

Reply:
done